README.md
# Sofa Image Resizer Service [![Build Status](https://travis-ci.org/sofa/sofa-image-resizer-service.png?branch=master)](https://travis-ci.org/sofa/sofa-image-resizer-service)

This is the Sofa Image Resizer Service component. 

## Installation

Via npm:

```sh
$ npm install sofa-image-resizer-service
```

Via Bower:

```sh
$ bower install sofa-image-resizer-service
```


CHANGELOG.md
<a name="0.2.1"></a>
### 0.2.1 (2014-07-04)


#### Bug Fixes

* **sofa-image-resize-service:** image url encoding ([39939298](https://github.com/sofa/sofa-image-resizer-service/commit/39939298608c7bc4c77c7b2d486930f3259b77e6))


<a name="0.2.1"></a>
### 0.2.1 (2014-07-04)


#### Bug Fixes

* **sofa-image-resize-service:** image url encoding ([39939298](https://github.com/sofa/sofa-image-resizer-service/commit/39939298608c7bc4c77c7b2d486930f3259b77e6))


<a name="0.2.0"></a>
## 0.2.0 (2014-06-23)


#### Bug Fixes

* **gitignore:** add bower_components ([2a9f1e13](https://github.com/sofa/sofa-image-resizer-service/commit/2a9f1e137c5249a6a6b695c794f767d3bafce0c7))


#### Features

* **grunt:** better deploying ([8d0328a5](https://github.com/sofa/sofa-image-resizer-service/commit/8d0328a50f4c31961913212c80aead21e14841ed))


<a name="0.1.1"></a>
### 0.1.1 (2014-06-18)


#### Bug Fixes

* **sofa-image-resizer-service:** brought back the base64encode polyfill ([297f42fa](https://github.com/sofa/sofa-image-resizer-service/commit/297f42fa7c9731405c6e43f0c6fae7f9e755157f))


<a name="0.1.0"></a>
## 0.1.0 (2014-04-16)


#### Bug Fixes

* **Gruntfile:** fixes license banner ([f86a8f27](https://github.com/sofa/sofa-image-resizer-service/commit/f86a8f27c6c77236e822451789c089af172ec516))
* **LICENSE:** fixes naming ([8fc0fb21](https://github.com/sofa/sofa-image-resizer-service/commit/8fc0fb21f48c4410266ace55654c0ae5fd93cdfd))



test/sofa.imageResizerService.spec.js
'use strict';
/* global sofa */

describe('sofa.ImageResizerService', function () {

    var imageResizerService,
        configService;

    beforeEach(function () {
        configService = new sofa.ConfigService();
        imageResizerService = new sofa.ImageResizerService(configService, window);
    });

    it('should be defined', function () {
        expect(imageResizerService).toBeDefined();
    });

    it('should have a method resize', function () {
        expect(imageResizerService.resize).toBeDefined();
    });

    describe('sofa.ImageResizerService#resize', function () {

        it('should be a function', function () {
            expect(typeof imageResizerService.resize).toBe('function');
        });

        it('should require maxwidth/maxheight', function () {
            expect(function () {
                imageResizerService.resize();
            }).toThrow('maxwidth and maxheight are required parameters');
        });

        it('should return a url to the CDN', function () {
            var resizedUrl = imageResizerService.resize('https://www.couchcommerce.com/images/couch_brand.png', { maxwidth: 100, maxheight: 100});
            expect(typeof resizedUrl).toBe('string');
            expect(resizedUrl.indexOf(configService.get('imageResizer')) > -1).toBe(true);
        });

        it('should accept parameters in query string format', function () {
            var resizedUrl = imageResizerService.resize('https://www.couchcommerce.com/images/couch_brand.png', 'maxwidth=100&maxheight=100');
            expect(typeof resizedUrl).toBe('string');
            expect(resizedUrl.indexOf(configService.get('imageResizer')) > -1).toBe(true);
        });

        it('should not resize when disabled', function () {
            sofa.Config.imageResizerEnabled = false;
            var originalUrl = 'https://www.couchcommerce.com/images/couch_brand.png';
            var resizedUrl = imageResizerService.resize(originalUrl, { maxwidth: 100, maxheight: 100});
            expect(resizedUrl).toEqual(originalUrl);
        });
    });
});


Gruntfile.js
'use strict';
/* jshint camelcase: false */

module.exports = function (grunt) {

    require('load-grunt-tasks')(grunt);

    grunt.initConfig({

        pkg: grunt.file.readJSON('package.json'),
        meta: {
            banner:
                '/**\n' +
                ' * <%= pkg.name %> - v<%= pkg.version %> - <%= grunt.template.today("yyyy-mm-dd") %>\n' +
                ' * <%= pkg.homepage %>\n' +
                ' *\n' +
                ' * Copyright (c) 2014 CouchCommerce GmbH (http://www.couchcommerce.com / http://www.sofa.io) and other contributors\n' +
                ' * THIS SOFTWARE CONTAINS COMPONENTS OF THE SOFA.IO COUCHCOMMERCE SDK (WWW.SOFA.IO).\n' +
                ' * IT IS PROVIDED UNDER THE LICENSE TERMS OF THE ATTACHED LICENSE.TXT.\n' +
                ' */\n'
        },
        component_name: 'sofa.imageResizerService',
        component_sass_name: grunt.file.readJSON('bower.json').name,
        build_dir: 'dist',

        clean: {
            build: {
                src: '<%= build_dir %>'
            }
        },

        html2js: {
            src: {
                options: {
                    base: 'src'
                },
                src: ['src/**/*.tpl.html'],
                dest: '<%= build_dir %>/<%= component_name %>.templates.js',
                module: '<%= component_name %>.templates'
            }
        },

        jshint: {
            options: {
                jshintrc: '.jshintrc'
            },

            src: {
                files: {
                    src: ['src/**/*.js']
                }
            },

            test_unit: {
                files: {
                    src: ['test/**/*.spec.js']
                }
            },

            gruntfile: {
                files: {
                    src: ['Gruntfile.js']
                }
            }
        },

        sass: {
            build: {
                options: {
                    style: 'expanded'
                },
                files: {
                    '<%= build_dir %>/<%= component_sass_name %>.css': 'src/<%= component_sass_name %>.scss',
                    '<%= build_dir %>/<%= component_sass_name %>-default.css': 'src/<%= component_sass_name %>-default.scss'
                }
            }
        },

        copy: {
            build: {
                files: [
                    {
                        src: ['**/*.js'],
                        dest: '<%= build_dir %>/',
                        cwd: 'src',
                        expand: true
                    }
                ]
            }
        },

        karma: {
            options: {
                configFile: 'karma.conf.js'
            },
            unit: {
                port: 9101,
                background: true
            },
            continuous: {
                singleRun: true
            }
        },

        cssmin: {
            compile: {
                options: {
                    report: 'gzip'
                },
                files: {
                    '<%= build_dir %>/<%= component_sass_name %>.css': 'src/<%= component_sass_name %>.css',
                    '<%= build_dir %>/<%= component_sass_name %>-default.css': 'src/<%= component_sass_name %>-default.css'
                }
            }
        },

        ngmin: {
            compile: {
                files: [
                    {
                        src: ['**/*.js'],
                        cwd: '<%= build_dir %>',
                        dest: '<%= build_dir %>',
                        expand: true
                    }
                ]
            }
        },

        concat: {
            options: {
                banner: '<%= meta.banner %>'
            },
            compile_js: {
                src: [
                    'component.prefix',
                    'src/**/*.js',
                    '<%= html2js.src.dest %>',
                    'component.suffix'
                ],
                dest: '<%= build_dir %>/<%= component_name %>.js'
            },
        },

        uglify: {
            compile: {
                options: {
                    banner: '<%= meta.banner %>'
                },
                files: {
                    '<%= build_dir %>/<%= component_name %>.min.js': '<%= concat.compile_js.dest %>'
                }
            },
        },

        changelog: {
            options: {
                dest: 'CHANGELOG.md'
            }
        },

        delta: {

            gruntfile: {
                files: 'Gruntfile.js',
                tasks: ['jshint:gruntfile'],
                options: {
                    livereload: false
                }
            },

            jssrc: {
                files: [
                    'src/**/*.js'
                ],
                tasks: ['jshint:src', 'karma:unit:run', 'ngmin', 'concat:compile_js', 'uglify:compile']
            },

            tpls: {
                files: [
                    '<%= html2js.src.src %>'
                ],
                tasks: ['html2js', 'concat:compile_js', 'uglify:compile']
            },

            jsunit: {
                files: [
                    'test/**/*.spec.js'
                ],
                tasks: ['jshint:test_unit', 'karma:unit:run']
            }
        },

        deploy: {
            options: {
                versionFiles: [
                    'package.json',
                    'bower.json'
                ],
                preDeployFn: function (grunt, newVersion, done) {
                    // Needed to make uglify use the banner with the new version inside
                    grunt.config.set('pkg', grunt.file.readJSON('package.json'));

                    grunt.task.run([
                        'build',
                        'changelog'
                    ]);
                    done();
                },
                postDeployFn: function (grunt, newVersion, done) {
                    if (!grunt.option('soft')) {
                        var exec = require('child_process').exec;
                        exec('npm publish', null, function (err, stdout) {
                            console.log(stdout);
                            done();
                        });
                    }
                    else {
                        done();
                    }
                }
            }
        }
    });

    grunt.registerTask('default', 'build');
    grunt.renameTask('watch', 'delta');

    grunt.registerTask('watch', [
        'build',
        'karma:unit',
        'delta'
    ]);

    grunt.registerTask('test', [
        'jshint',
        'karma:continuous'
    ]);

    grunt.registerTask('build', [
        'clean',
        'jshint',
        'sass',
        'karma:continuous',
        'cssmin:compile',
        'ngmin',
        'concat:compile_js',
        'uglify:compile'
    ]);
};


karma.conf.js
module.exports = function (karma) {
    karma.set({
        /**
         * From where to look for files, starting with the location of this file.
         */
        basePath: '.',

        /**
         * This is the list of file patterns to load into the browser during testing.
         */
        files: [
            'node_modules/sofa-core/dist/sofa.core.js',
            'node_modules/sofa-testing/mocks/sofa.config.mock.js',
            'src/**/*.js',
            'test/**/*.spec.js'
        ],
        exclude: [],
        frameworks: ['jasmine'],
        plugins: [
            'karma-jasmine',
            'karma-firefox-launcher',
            'karma-chrome-launcher',
        ],
        preprocessors: {},

        /**
         * How to report, by default.
         */
        reporters: 'dots',

        /**
         * On which port should the browser connect, on which port is the test runner
         * operating, and what is the URL path for the browser to use.
         */
        port: 9018,
        runnerPort: 9101,
        urlRoot: '/',

        /**
         * Disable file watching by default.
         */
        autoWatch: false,

        /**
         * The list of browsers to launch to test on. This includes only "Firefox" by
         * default, but other browser names include:
         * Chrome, ChromeCanary, Firefox, Opera, Safari, PhantomJS
         *
         * Note that you can also use the executable name of the browser, like "chromium"
         * or "firefox", but that these vary based on your operating system.
         *
         * You may also leave this blank and manually navigate your browser to
         * http://localhost:9018/ when you're running tests. The window/tab can be left
         * open and the tests will automatically occur there during the build. This has
         * the aesthetic advantage of not launching a browser every time you save.
         */
        browsers: [
            process.env.TRAVIS ? 'Firefox' : 'Chrome'
        ]
    });
};


src/sofa.imageResizerService.js
'use strict';
/* global sofa, base64Encode */

sofa.define('sofa.ImageResizerService', function (configService, $window) {

    // http://phpjs.org/functions/base64_encode/
    /* jshint ignore:start */
    var base64Encode = function (data) {
        if (typeof $window.btoa === 'function') {
            return $window.btoa(data);
        }
        var b64 = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/=';
        var o1, o2, o3, h1, h2, h3, h4, bits, i = 0,
            ac = 0,
            enc = '',
            tmpArr = [];
        if (!data) {
            return data;
        }
        do { // pack three octets into four hexets
            o1 = data.charCodeAt(i++);
            o2 = data.charCodeAt(i++);
            o3 = data.charCodeAt(i++);
            bits = o1 << 16 | o2 << 8 | o3;
            h1 = bits >> 18 & 0x3f;
            h2 = bits >> 12 & 0x3f;
            h3 = bits >> 6 & 0x3f;
            h4 = bits & 0x3f;
            // use hexets to index into b64, and append result to encoded string
            tmpArr[ac++] = b64.charAt(h1) + b64.charAt(h2) + b64.charAt(h3) + b64.charAt(h4);
        } while (i < data.length);
        enc = tmpArr.join('');
        var r = data.length % 3;
        return (r ? enc.slice(0, r - 3) : enc) + '==='.slice(r || 3);
    };
    /* jshint ignore:end */

    //http://stackoverflow.com/questions/1714786/querystring-encoding-of-a-javascript-object
    function objectToQueryString(obj) {
        var str = [];
        for (var p in obj) {
            if (obj.hasOwnProperty(p)) {
                str.push(encodeURIComponent(p) + '=' + encodeURIComponent(obj[p]));
            }
        }
        return str.join('&');
    }

    // http://www.developerdrive.com/2013/08/turning-the-querystring-into-a-json-object-using-javascript/
    function queryStringToObject(queryString) {
        var pairs = queryString.split('&');
        var result = {};
        pairs.forEach(function (pair) {
            pair = pair.split('=');
            result[pair[0]] = decodeURIComponent(pair[1] || '');
        });
        return JSON.parse(JSON.stringify(result));
    }

    var resizeUrlCache = {},
        self = {};

    self.resize = function (imageUrl, args) {
        if (!configService.get('imageResizerEnabled')) {
            return (imageUrl);
        }

        var cacheKey;
        if (sofa.Util.isString(args)) {
            cacheKey = imageUrl + args;
        } else {
            cacheKey = imageUrl + objectToQueryString(args);
        }
        if (resizeUrlCache[cacheKey]) {
            return resizeUrlCache[cacheKey];
        }

        // no cache hit, decode string argument to object if required
        if (sofa.Util.isString(args)) {
            args = queryStringToObject(args);
        }
        if (!args || !args.hasOwnProperty('maxwidth') || !args.hasOwnProperty('maxheight')) {
            throw new Error('maxwidth and maxheight are required parameters');
        }

        // Add protocol if not specified in URL
        if (/^\/\//.test(imageUrl)) {
            imageUrl = $window.location.protocol + imageUrl;
        }

        var defaults = {
            cmd: 'resize',
            url: imageUrl,
            quality: 100,
            devicePixelRatio: $window.devicePixelRatio
        },
            fullArgs = sofa.Util.extend(args, defaults),
            imageExt = imageUrl.substring(imageUrl.lastIndexOf('.') + 1);

        // In case of retina display, need to fetch larger image
        if (args.devicePixelRatio > 1) {
            fullArgs.maxwidth *= args.devicePixelRatio;
            fullArgs.maxheight *= args.devicePixelRatio;
        }

        var encodedCall = base64Encode(objectToQueryString(fullArgs)).replace('/', '_') + '.' + imageExt,
            resizedImageUrl = configService.get('imageResizerEndpoint') + encodedCall;

        resizeUrlCache[cacheKey] = resizedImageUrl;

        return resizedImageUrl;
    };

    return self;
});
